File: 2021/12-passage-pathing/passage-pathing.py
from string import ascii_uppercase as UC, ascii_lowercase as LC
from collections import defaultdict

TASK = 2 # 
A, visited = defaultdict(list), defaultdict(int)

def dfs(v, twice):
    if v=='end': return 1
    visited[v] += 1
    paths = 0
    for u in A[v]:
        if u[0] in UC or u[0] in LC and not visited[u]:
            paths += dfs(u, twice)
        elif TASK==2 and u != 'start' and not twice:
            paths += dfs(u, twice=True)
    visited[v] -= 1
    return paths

for line in open(0):
    u, v = line.rstrip().split('-')
    A[u].append(v)
    A[v].append(u)

print(dfs('start', twice=False))

File: 2021/12-passage-pathing/passage-pathing.go
// start: 2021-12-12T07:10:27Z

package main

import (
    "fmt"
    "bufio"
    "os"
    . "strings"
    "sort"
)

const (
	UC = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
	TASK = 2
)

func dfs(A map[string][]string, v string, visited map[string]int, twice bool) int {
	if v == "end" {
		return 1
	}
	visited[v] = visited[v] + 1
	paths := 0	
	for _, u := range A[v] {
		if Contains(UC, string(u[0])) || visited[u] == 0 {
			paths += dfs(A, u, visited, twice)
		} else {
			if TASK==2 && u != "start" && !twice {
				paths += dfs(A, u, visited, true)
			}
		}
	}
	visited[v]--
	return paths
}

func main() {
	reader := bufio.NewReader(os.Stdin)
    scn := bufio.NewScanner(reader)
	A := make(map[string][]string)
    for scn.Scan() {
    	pair := SplitN(scn.Text(), "-", -1)
    	for i:=0; i<2; i++ {
    		u, v := pair[i], pair[1-i]
	    	neighbors, exists := A[u]
	    	if exists {
	    		A[u] = append(neighbors, v)
	    	} else {
	    		A[u] = make([]string, 1)
	    		A[u][0] = v
	    	}
	    }
    }
    for v, _ := range A {
    	sort.Strings(A[v])
    }
    fmt.Println(dfs(A, "start", make(map[string]int), false))

// end: 2021-12-12T15:03:59Z stupid mistake left out some letters from UC!!!>:\